### ISSUE TITLE
Add a "Try me" Python REPL to Hiperwalk website 

### ISSUE THREAD

**Issue by gmgall** [hiperwalk/hiperwalk.github.io#4]
It would be nice to have a REPL like the one on https://numpy.org/ for potential users to test the package without installing it.

![Screenshot 2023-10-03 at 18-56-17 NumPy](https://github.com/hiperwalk/hiperwalk.github.io/assets/973834/ee428289-7e5b-4226-8795-bd5e7436e520)

**Comment by gmgall** [comment]
I've been studying how https://numpy.org/ does it and I will report what I discovered bellow.

# Layout wise

[scientific-python-hugo-theme](https://github.com/scientific-python/scientific-python-hugo-theme) uses a [partial](https://gohugo.io/templates/partials/) called `shell.html` if it exists. See https://github.com/scientific-python/scientific-python-hugo-theme/blob/main/layouts/index.html#L22

We need to create a section in the `config.yaml` file and the `partial`.

Adding a section below `params` like

```yaml
  shell:
    intro:
    - title: Try Hiperwalk
      text: Use the interactive shell to try Hiperwalk in the browser
```

in `config.yaml` and creating a `layouts/partials/shell.html` file with the following content

```html
{{- $shell := .Site.Params.shell }}
{{- $intro := index $shell "intro" }}
<div class="hero-right">
    <div class="flex-column shell-title-container">
        {{- range $intro }}
            <div class="shell-title">{{ .title }}</div>
            <div class="shell-content-message">
                <p>{{ .text }}</p>
            </div>
        {{- end }}
    </div>
    <div class="numpy-shell-canvas">
        <div class="numpy-shell-container">
            <div class="shell-lesson shell-content">
                {{partial "shell-lesson.html" | print | markdownify}}
            </div>
            <!-- Interactive Shell -->
            <iframe
                class="numpy-shell"
                src="https://gmgall.github.io/xeus-python/repl/?toolbar=1&kernel=python&code=import%20hiperwalk%20as%20hw"
            >
            </iframe>
        </div>
    </div>
</div>
```

would suffice.

# REPL wise

The interactive shell in https://numpy.org/ is a [JupyterLite](https://github.com/jupyterlite/jupyterlite) REPL loaded via an [<iframe>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe). JupyterLite is Jupyter powered by [Wasm](https://developer.mozilla.org/en-US/docs/WebAssembly) i.e. it runs entirely on the client's browser.

NumPy's website simply does

```html
<iframe
  class="numpy-shell"
  src="https://jupyterlite.github.io/demo/repl/?toolbar=1&amp;kernel=python&amp;code=import%20numpy%20as%20np">
</iframe>
```

and has a functional Python REPL with NumPy installed. This works "out of the box" for them because the demo on https://jupyterlite.github.io/demo/repl/ uses a [Pyodide](https://pyodide.org/en/stable/) kernel. Pyodide brings several Python packages already installed. NumPy is one of them.

Hiperwalk would need to be installed separately.

Under [Choosing a kernel](https://jupyterlite.readthedocs.io/en/latest/howto/configure/kernels.html#choosing-a-kernel) in the JupyterLite docs is described that 2 Python kernels exist.

Quoting the documentation:

## [`jupyterlite-pyodide-kernel`](https://jupyterlite.readthedocs.io/en/latest/howto/configure/kernels.html#jupyterlite-pyodide-kernel)

* Is based on [Pyodide](https://github.com/pyodide/pyodide)
* Uses [IPython](https://github.com/ipython/ipython) for the code execution (access to IPython magics, support for the inline Matplotlib backend, etc)
* Provides a way to dynamically install packages with piplite (e.g. await piplite.install("ipywidgets"))
*  Does not support sleeping with from time import sleep
*  **Does not support pre-installing packages**

## [`jupyterlite-xeus-python`](https://jupyterlite.readthedocs.io/en/latest/howto/configure/kernels.html#jupyterlite-pyodide-kernel)

* Is based on [xeus-python](https://github.com/jupyter-xeus/xeus-python)
* Uses [IPython](https://github.com/ipython/ipython) for the code execution (access to IPython magics, support for the inline Matplotlib backend, etc)
* Does not provide (yet) a way to dynamically install packages. There is ongoing work for building a mamba package manager for WebAssembly that would allow for installing packages on the fly.
* Supports sleeping with from time import sleep
* **Supports pre-installing packages** from emscripten-forge and conda-forge, by providing an environment.yml file defining the runtime environment

Since we need a kernel with Hiperwalk preinstalled, I believe `jupyterlite-xeus-python` is our way to go.

I managed to generate a `xeus-python` kernel with hyperwalk preinstalled after some attempts, but I couldn't import it. 

```python
ModuleNotFoundError: No module named 'gi'
```

This `import` error happens because I wasn't able to install the `PyGObject` package. It became a dependency after hiperwalk/hiperwalk@358fc53a03e5051dee1c4ebe8ac3d0f13d6c8692.

To try my attempt of creating a xeus-python kernel with hiperwalk installed: https://gmgall.github.io/xeus-python/repl/?toolbar=1&kernel=python&code=import%20hiperwalk%20as%20hpw

**Comment by gmgall** [comment]
BTW: hiperwalk was installed in the kernel via `pip`.

This is experimental [according to xeus-python docs](https://xeus-python-kernel.readthedocs.io/en/latest/configuration.html#pip-packages).

@gustavowl, is Hiperwalk available available through conda/mamba?

**Comment by gustavowl** [comment]
For now, hiperwalk is only available via `pip`. I have to check if there exists an automatic pip-conda integration.

The first attempt looks promising. I would like to open discussion on a couple of topics.

## Suppressing warnings ##

Should we suppress the warning when the user machine has no hpc support?

- By doing so, some warnings will not appear on the REPL. Or could we just suppress them on the REPL?
- How would a user know that the hpc installation was unsuccessful?

The discussion may be rather extensive.
I may open an issue on the topic.

## `gi` error handling ##

The `gi` error seems natural. After all, `gi` is only used for plotting on gnome.
It is related to a quick solution to [issue #6](https://github.com/hiperwalk/hiperwalk/issues/6#issuecomment-1682977145).
Basically `gi` considered as a dependency when it actually shouldn't.
But importing `gi` on demand would not show animations.

Is it possible for the REPL to have image and animation support?
Or should we have another webpage dedicated to it?

**Comment by gmgall** [comment]
I built a JupyterLite website installing `hiperwalk` via `pip` and everything seems to be working fine. It looks like we don't need to bother with making `hiperwalk` available via mamba for now.

---

About _Suppressing warnings_: I think this deserve a more careful deliberation. I'll open an issue about it.

---

About _`gi` error handling_: the `gi` package isn't used anymore.

**Comment by gmgall** [comment]
After hiperwalk/hiperwalk#32, I'm able to build a JupyterLite site with hiperwalk available. The animations work as well.

To avoid linking to my personal account I still need the permissions to create a repository in the hiperwalk account to send a PR to add the REPL to https://hiperwalk.org.

Some apps of the JupyterLite site:

- the [REPL](https://gmgall.github.io/xeus-python/repl/?toolbar=1&kernel=xeus-python&code=import%20hiperwalk%20as%20hpw) that will be embed on hiperwalk.org
- [Lab](https://gmgall.github.io/xeus-python/lab/) interface
- [Notebook](https://gmgall.github.io/xeus-python/retro/notebooks/?path=Tutorial.ipynb) interface

**Comment by gmgall** [comment]
:ok: A JupyterLite site with hiperwalk available is now being built in [hiperwalk/xeus-python](https://github.com/hiperwalk/xeus-python).

A REPL like the one on https://numpy.org is available via https://hiperwalk.github.io/xeus-python/repl/?toolbar=1&kernel=xeus-python&code=import%20hiperwalk%20as%20hpw

---

:warning: We need to decide now which code example we are going to list on the left of the REPL.

![hiperwalk org_repl](https://github.com/hiperwalk/hiperwalk.github.io/assets/973834/3f20f9fe-f571-4db2-84f7-1c3281bc7c1d)

**Comment by gmgall** [comment]
85f63242a4c7adee78b3e3f2cd9730461b21e498 added a 'Try Hiperwalk' section with a Python REPL.
